### DISCUSSION TITLE
indexNode、queryNode和dataNode

### DISCUSSION THREAD

**Discussion by RongHouchao** [milvus-io/milvus#20802] [Q&A and General discussion] [answered]
indexNode、queryNode和dataNode 扩容哪个node会对向量近似召回 性能有明显的提升？
indexNode只是在存储向量构建索引的时候起作用吗？向量召回的过程他起什么作用？
每个dataNode都会加载全量的向量数据吗？

**Comment by yhmo** [comment, marked as answer]
- queryNode is mainly for ANN search, query node loads segments into memory for search. Increasing the number of query nodes can get higher search performance, but not always. Some affect factors: number of segments, segment size, replica number. Assume you have a collection with 2 segments, then you call collection.load(replica=1), increase the query nodes from 1 to 2, you can get better performance. But if you continue to increase query nodes from 2 to 3, the search performance is no change because there are only 2 segments(one segment per query node).

- dataNode is mainly for data ingestion and storage. But with our experience, increasing the number of data node affect insert speed slightly. The data node only processes insert/delete operations, it doesn't load segments into memory.

- indexNode is mainly for building index. Each index node builds index for one segment each time. With more index nodes, they can build index parallelly for more segments.

**Reply by RongHouchao** [reply]
所以，queryNode的数量影响的是向量search性能，indexNode和dataNode的数量影响数据写入的性能。
indexNode和dataNode的数量扩容只会影响数据写入的效率，但对向量search性能没有任何影响。
是这样的吧？

**Reply by yhmo** [reply]
> 所以，queryNode的数量影响的是向量search性能，indexNode和dataNode的数量影响数据写入的性能。 indexNode和dataNode的数量扩容只会影响数据写入的效率，但对向量search性能没有任何影响。 是这样的吧？

indexNode所建的索引是为search服务的，没来得及建索引的数据是暴搜的，如果索引建得快，search就能更早地用上索引，从这点来说，indexNode对search性能是有影响的。
